src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java

```diff
@@ -167,163 +157,203 @@ private static int getPositionAfter(DetailAST ast) {
+    public void setAllowLineBreaks(boolean allowLineBreaks) {
+        this.allowLineBreaks = allowLineBreaks;
     }
 
     /**
-     * Checks if current array is multidimensional.
-     * @param arrayDeclaration {@link TokenTypes#ARRAY_DECLARATOR ARRAY_DECLARATOR}
-     * @return true if current array is multidimensional.
+     * Checks if current line has redundant whitespace after specified index.
+     * @param line line of java source.
+     * @param after specified index.
+     * @return true if line contains redundant whitespace.
      */
-    private static boolean isMultiDimensionalArray(DetailAST arrayDeclaration) {
-        return arrayDeclaration.getParent().getType() == TokenTypes.ARRAY_DECLARATOR
-                || arrayDeclaration.getFirstChild().getType() == TokenTypes.ARRAY_DECLARATOR;
+    private boolean hasRedundantWhitespace(String line, int after) {
+        boolean result = !allowLineBreaks;
+        for (int i = after + 1; !result && i < line.length(); i++) {
+            if (!Character.isWhitespace(line.charAt(i))) {
+                result = true;
+            }
+        }
+        return result;
     }
 
     /**
-     * Checks if current array declaration is part of array instantiation.
-     * @param arrayDeclaration {@link TokenTypes#ARRAY_DECLARATOR ARRAY_DECLARATOR}
-     * @return true if current array declaration is part of array instantiation.
+     * Gets previous node for {@link TokenTypes#INDEX_OP INDEX_OP} token
+     * for usage in getPositionAfter method, it is a simplified copy of
+     * getArrayDeclaratorPreviousElement method.
+     * @param ast
+     *        , {@link TokenTypes#INDEX_OP INDEX_OP} node.
+     * @return previous node by text order.
      */
-    private static boolean isArrayInstantiation(DetailAST arrayDeclaration) {
-        return arrayDeclaration.getParent().getType() == TokenTypes.LITERAL_NEW;
+    private static DetailAST getIndexOpPreviousElement(DetailAST ast) {
+        final DetailAST firstChild = ast.getFirstChild();
+        if (firstChild.getType() == TokenTypes.INDEX_OP) {
+            // second or higher array index
+            return firstChild.findFirstToken(TokenTypes.RBRACK);
+        }
+        else {
+            // always has an ident token
+            return getIdentLastToken(ast);
+        }
     }
 
     /**
-     * Checks if current array is used as type for generic bounded wildcard.
-     * <p>
-     * E.g. {@code <? extends String[]>} or {@code <? super Object[]>}.
-     * </p>
-     * @param arrayDeclarator {@link TokenTypes#ARRAY_DECLARATOR ARRAY_DECLARATOR}
-     * @return true if current array is used as type for generic bounded wildcard.
+     * Returns proper argument for getPositionAfter method, it is a token after
+     * {@link TokenTypes#ARRAY_DECLARATOR ARRAY_DECLARATOR}, in can be {@link TokenTypes#RBRACK
+     * RBRACK}, {@link TokenTypes#INDENT INDENT} or an array type definition (literal).
```

Comment:
IDEA Inspections have just detected problem in this line - there's no `TokenTypes.INDENT` available: https://teamcity.jetbrains.com/viewLog.html?buildId=563116&tab=Inspection&buildTypeId=Checkstyle_IdeaInspectionsPullRequest